In [1]:
import duckdb
import polars as pl
from polars import col as c
import polars.selectors as cs
from config import BASE_PARQUET_PATH, DATABASE_URL


## Hospital Analysis

Using the hospital data stored in the `hospital_directory` table of the duckdb database, perform the following analyses:

1. Determine the total number of hospitals processed `processed`
2. Determine the number of hospitals with files compliant with CMS requirements `compliant`
3. Determine the number of hospitals with drugs listed in the file `has_drugs`
4. Determine the number of hospitals with drugs and corresponding prices `has_prices`

In [20]:
def run_query(query: str) -> pl.LazyFrame:
    with duckdb.connect(DATABASE_URL) as con:
        return pl.DataFrame(con.execute(query=query).fetch_arrow_table()).lazy()

def base_query():
    sql = f"""select * from hospital_directory"""
    return run_query(sql)


In [21]:
(
base_query()
.select(cs.boolean().exclude('converted').sum())
.collect(engine='streaming')
.to_pandas()
)

processed  compliant  has_drugs  has_pricing
0       2086       1697       1378          634

In [22]:
(
base_query()
.select(cs.boolean().exclude('converted').sum())
.with_columns((pl.all() / c.processed).round(4).name.suffix('_pct'))
.select(cs.matches('.*_pct').exclude('processed_pct'))
.collect(engine='streaming')
.to_pandas()
)

compliant_pct  has_drugs_pct  has_pricing_pct
0         0.8135         0.6606           0.3039

### Overall Summary Statistics
- 2,086 files were processed of which 81.35% were compliant with CMS requirements. 66.06% of the files had drugs listed and 30.39% had both drugs and corresponding prices.

## State based Summary Statistics

In [23]:
(
base_query()
.group_by("state")
.agg(cs.boolean().exclude('converted').sum())
.sort("state")
.collect(engine='streaming')
.to_pandas()
)

state  processed  compliant  has_drugs  has_pricing
0     AK         26         17         14            4
1     AL         43         39         26            8
2     AR         46         39         30           22
3     AZ         43         36         17           12
4     CA         50         41         40           21
5     CO         45         33         31           16
6     CT         33         22         12            1
7     DE         15         13          7            2
8     FL         50         35         25           12
9     GA         47         43         36           13
10    HI         25         24         21           12
11    IA         47         29         25            7
12    ID         45         33         28           18
13    IL         51         38         36           11
14    IN         52         40         30            8
15    KS         53         39         35           14
16    KY         45         33         22            7
17    LA         52         35         24           14
18    MA         47         38         24           10
19    MD         37         37         29            7
20    ME         33         31         28           15
21    MI         47         43         33           10
22    MN         46         44         42            8
23    MO         44         37         31           15
24    MS         56         33         28           18
25    MT         48         36         34           11
26    NC         44         37         36           30
27    ND         37         32         28           10
28    NE         51         39         39           29
29    NH         30         28         26           13
30    NJ         43         37         25            7
31    NM         43         38         30           17
32    NV         45         38         32           22
33    NY         55         34         27           12
34    OH         48         43         34           14
35    OK         52         38         29            6
36    OR         44         38         37           11
37    PA         48         43         34           18
38    RI         14         11          9            2
39    SC         41         39         28           14
40    SD         42         40         37           20
41    TN         47         43         29           23
42    TX         30         26         16            7
43    UT         28         23         17           15
44    VA         48         41         33           18
45    VT         13         11         10            9
46    WA         51         35         33           12
47    WI         43         39         34            2
48    WV         40         38         31           22
49    WY         23         18         16            5

In [24]:
(
base_query()
.group_by("state")
.agg(cs.boolean().exclude('converted').sum())
.with_columns((cs.matches('(?i)comp|has') / c.processed).round(4).name.suffix('_pct'))
.select(cs.matches('state|.*_pct').exclude('processed_pct'))
.sort("state")
.collect(engine='streaming')
.to_pandas()
)

state  compliant_pct  has_drugs_pct  has_pricing_pct
0     AK         0.6538         0.5385           0.1538
1     AL         0.9070         0.6047           0.1860
2     AR         0.8478         0.6522           0.4783
3     AZ         0.8372         0.3953           0.2791
4     CA         0.8200         0.8000           0.4200
5     CO         0.7333         0.6889           0.3556
6     CT         0.6667         0.3636           0.0303
7     DE         0.8667         0.4667           0.1333
8     FL         0.7000         0.5000           0.2400
9     GA         0.9149         0.7660           0.2766
10    HI         0.9600         0.8400           0.4800
11    IA         0.6170         0.5319           0.1489
12    ID         0.7333         0.6222           0.4000
13    IL         0.7451         0.7059           0.2157
14    IN         0.7692         0.5769           0.1538
15    KS         0.7358         0.6604           0.2642
16    KY         0.7333         0.4889           0.1556
17    LA         0.6731         0.4615           0.2692
18    MA         0.8085         0.5106           0.2128
19    MD         1.0000         0.7838           0.1892
20    ME         0.9394         0.8485           0.4545
21    MI         0.9149         0.7021           0.2128
22    MN         0.9565         0.9130           0.1739
23    MO         0.8409         0.7045           0.3409
24    MS         0.5893         0.5000           0.3214
25    MT         0.7500         0.7083           0.2292
26    NC         0.8409         0.8182           0.6818
27    ND         0.8649         0.7568           0.2703
28    NE         0.7647         0.7647           0.5686
29    NH         0.9333         0.8667           0.4333
30    NJ         0.8605         0.5814           0.1628
31    NM         0.8837         0.6977           0.3953
32    NV         0.8444         0.7111           0.4889
33    NY         0.6182         0.4909           0.2182
34    OH         0.8958         0.7083           0.2917
35    OK         0.7308         0.5577           0.1154
36    OR         0.8636         0.8409           0.2500
37    PA         0.8958         0.7083           0.3750
38    RI         0.7857         0.6429           0.1429
39    SC         0.9512         0.6829           0.3415
40    SD         0.9524         0.8810           0.4762
41    TN         0.9149         0.6170           0.4894
42    TX         0.8667         0.5333           0.2333
43    UT         0.8214         0.6071           0.5357
44    VA         0.8542         0.6875           0.3750
45    VT         0.8462         0.7692           0.6923
46    WA         0.6863         0.6471           0.2353
47    WI         0.9070         0.7907           0.0465
48    WV         0.9500         0.7750           0.5500
49    WY         0.7826         0.6957           0.2174